lib/galaxy/model/__init__.py

```diff
@@ -2440,15 +2441,33 @@ def add_pending_items(self, set_output_hid=True):
         self._pending_additions = []
 
     def _next_hid(self, n=1):
-        # this is overriden in mapping.py db_next_hid() method
-        if len(self.datasets) == 0:
-            return n
-        else:
-            last_hid = 0
-            for dataset in self.datasets:
-                if dataset.hid > last_hid:
-                    last_hid = dataset.hid
-            return last_hid + n
+        """
+        Generate next_hid from the database in a concurrency safe way:
+        1. Retrieve hid_counter from database
+        2. Increment hid_counter by n and store in database
+        3. Return retrieved hid_counter.
+
+        Handle with SQLAlchemy Core to keep this independent from current session state, except:
+        expire hid_counter attribute, since its value in the session is no longer valid.
+        """
+        session = object_session(self)
+        engine = session.bind
+        table = self.__table__
+        history_id = cached_id(self)
+        update_stmt = update(table).where(table.c.id == history_id).values(hid_counter=table.c.hid_counter + n)
+
+        with engine.begin() as conn:
```

Comment:
That creates a new commit, which we don't want. Why is this needed ?

Reply:
item 2 in the PR description. In my understanding, there's no autocommit in SA2.0. i.e., it would work now using `with endine.connect() as conn`, but unless we did `conn.commit()` in the end, it wouldn't work as soon as we moved to SA2.0.

Reply:
Can you verify we're not expiring all attributes here ? If we are this would be a huge performance penalty. It seems we might not be, since you're explicitly exping hid_counter ?

Reply:
I added a test for that very reason :smiley:  4ea118584da2bcc2e4be0080abdac56032308faa